src/network/eventloop.rs

```diff
@@ -44,27 +62,49 @@ impl EventLoop {
         }
     }
 
-    pub async fn run(mut self) {
+    /// Loop and select over swarm and pubsub [events] and client [commands].
+    ///
+    /// [events]: SwarmEvent
+    /// [commands]: Command
+    pub async fn run(mut self) -> Result<()> {
         loop {
             tokio::select! {
-                event = self.swarm.next() => self.handle_event(event.expect("Swarm stream to be infinite.")).await  ,
-                command = self.command_receiver.recv() => match command {
-                    Some(c) => self.handle_command(c).await,
-                    // Command channel closed, thus shutting down the network event loop.
-                    None=>  return,
-                },
+                event = self.swarm.select_next_some() => self.handle_event(event).await,
+                command = self.command_receiver.recv() => if let Some(c) = command {self.handle_command(c).await}
             }
         }
     }
 
-    async fn handle_event(
-        &mut self,
-        event: SwarmEvent<
-            ComposedEvent,
-            EitherError<ConnectionHandlerUpgrErr<io::Error>, io::Error>,
-        >,
-    ) {
+    async fn handle_event(&mut self, event: SwarmEvent<ComposedEvent, HandlerErr>) {
         match event {
+            SwarmEvent::Behaviour(ComposedEvent::Floodsub(FloodsubEvent::Message(message))) => {
```

Comment:
Why support FloodSub?

Reply:
There are some benefits if we want a direct pubsub vs a mesh w/ gossipsub. This is mainly for testing ground purposes right now, as there were some interesting mesh issues I ran into before getting things to work concretely. Once we're ready for a real "version" of things, we can probably cut it, but it's good for experimentation.